Package WinApps (and the Launcher) with Nix

oskardotglobal opened the pull request:
This depends on #232, but this PR makes WinApps and the launcher available to install conveniently using the Nix package manager.

I'll look into getting this into Nixpkgs, the Nix package registry, but this might not be possible since WinApps is technically licensed under ARR.

oskardotglobal committed:
feat: package winapps & the winapps launcher using nix

doc: document installation using Nix

fix: explicitely specify path to inquirer

fix: libvirt group is called libvirtd on nix

fix: set libvirturl for winapps-launcher

fix: refine group change sed expr

feat: add desktop file

fix: do not use source derivation

fix: use install instead of cp

fix: ensure  exists

LDprg commented:
Great fan of this, since I use nixos myself.

AkechiShiro commented:
Once I do test #232, I can give this a try, thanks a lot for the work @oskardotglobal

oskardotglobal commented:
When #15 gets merged, I'll also try to get this into nixpkgs.

oskardotglobal committed:
fix: do not symlink binaries on nix

oskardotglobal committed:
feat: bump version

oskardotglobal committed:
Merge branch 'main' into feat-nix-packaging

oskardotglobal committed:
fix: properly escape ${}

oskardotglobal committed:
feat: bump version

AkechiShiro commented on packages/winapps/default.nix:
Can you nixfmt the file ?

oskardotglobal replied:
It is formatted using `alejandra`

AkechiShiro replied:
Hm, could we use the official nixfmt, there was an RFC that was worked by the nix community, if you want to use something else locally, that's fine, but I don't think we should roll our own format of Nix code, but then you'll see that it will fail to land upstream as there are checks for the format now.

AkechiShiro commented on packages/winapps/default.nix:
```diff
@@ -42,8 +42,6 @@ in
 
       sed -E -i \
```
I'm not sure sed usage is the best to do when patching a package, I believe there are specific fonctions for this called `substituteInPlace` , here is some discussion about when to use it or not : https://discourse.nixos.org/t/when-to-use-substituteinplace-functions-vs-a-patch/11073

oskardotglobal replied:
I've just tried swapping all the calls to `sed` with `substituteInPlace`, but now the patterns do not match anymore

AkechiShiro replied:
Maybe upstream will have better insights than me, I would recommend you to generate a patch with all the sed commands and apply the patch instead of having all these complex escaped sed in the source derivation.

AkechiShiro commented on packages/winapps/default.nix:
```diff
@@ -0,0 +1,75 @@
+      description = "Run Windows applications (including Microsoft 365 and Adobe Creative Cloud) on GNU/Linux with KDE, GNOME or XFCE, integrated seamlessly as if they were native to the OS. Wayland is currently unsupported.";
+      mainProgram = "winapps";
+      platforms = platforms.linux;
+      license = licenses.gpl3;
```
You'll need to change this to `agpl3plus`

AkechiShiro commented on packages/winapps/default.nix:
```diff
@@ -0,0 +1,75 @@
+
+      cp -r ./ $out/src/
+
+      install -m755 -D bin/winapps $out/bin/winapps
```
Is it possible to use `555` for the permission ? or is 755 needed ?

oskardotglobal replied:
The store is mounted as readonly anyways and we need the 755 to be able to patch it, so I think it's fine

oskardotglobal committed:
feat: use nixfmt

oskardotglobal committed:
feat: use patches

oskardotglobal committed:
fix: properly name patches & remove nixfmt pre-commit hook since it requires GHC & hackage

AkechiShiro commented:
@oskardotglobal can you PR the package upstream on Nixpkgs as a draft PR maybe ? Or if you did test already, PR it for reviews

oskardotglobal commented:
Yeah, I will

oskardotglobal commented:
It's just currently not building

oskardotglobal committed:
Merge remote-tracking branch 'origin/main' into feat-nix-packaging

oskardotglobal committed:
fix: fix malformed patch

oskardotglobal commented:
Alright, this builds properly again; it would be good if someone were to try this too
I use it on my system but I can't catch all the bugs

AkechiShiro commented:
@oskardotglobal Will let you know once I tried, can you tell me a way to try it ? Should I drop it into a nixpkgs folder and build it then use it ?

I need to setup a Windows VM with `quickemu` before using it right ?

oskardotglobal commented:
@AkechiShiro No, you will need either a VM using `libvirt`, `docker` or `podman`. Just follow the normal installation guide in the README. There are also instructions on how to install winapps via the flake in this repository. You can also check my .dotfiles: <https://github.com/oskardotglobal/.dotfiles/blob/nix/flake.nix>

AkechiShiro commented:
Couldn't try the flake as it is not in the repo yet, not sure if I can specify the branch inside the flake syntax here : 
```
nix profile install github:winapps-org/winapps#winapps
error: getting status of '/nix/store/341r1hwprc108jk6cf7c0ryxblwwp7dp-source/flake.nix': No such file or directory
```

oskardotglobal commented:
You have to add `/feat-nix-packaging` to the end of the flake input definition or add a commit hash from that branch

AkechiShiro commented:
Step 3 needs to be done, I think we need more documentation on how to use it, I believe `winapps-setup` must be run after step 3 is done and then `winapps` or `winapps-launcher` can be used.

AkechiShiro commented:
MyRDP connection seems to fail, not sure why exactly, I setup W11 using Docker and used the USERNAME and PASSWORD set.
```
[21:33:11:432] [289627:00046b5c] [ERROR][com.freerdp.core.transport] - [transport_read_layer]: BIO_read returned a system error 104: Connection reset by peer
[21:33:11:432] [289627:00046b5c] [ERROR][com.freerdp.core] - [transport_read_layer]: ERRCONNECT_CONNECT_TRANSPORT_FAILED [0x0002000D]
```

AkechiShiro commented:
Here is the detailed logs using `WLOG_LEVEL=DEBUG`: 
```
[21:36:46:998] [290759:00046fc8] [DEBUG][com.freerdp.core] - [freerdp_connect_begin]: resetting error state
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.client.common.cmdline] - [freerdp_client_load_static_channel_addin]: loading channelEx rdpdr
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.client.common.cmdline] - [freerdp_client_load_static_channel_addin]: loading channelEx rdpsnd
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.channels.channels.cliprdr.client] - [cliprdr_VirtualChannelEntryEx]: VirtualChannelEntryEx
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.client.common.cmdline] - [freerdp_client_load_static_channel_addin]: loading channelEx cliprdr
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.channels.drdynvc.client] - [drdynvc_VirtualChannelEntryEx]: VirtualChannelEntryEx
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.client.common.cmdline] - [freerdp_client_load_static_channel_addin]: loading channelEx drdynvc
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.primitives] - [primitives_autodetect_best]: primitives benchmark: only one backend, skipping...
[21:36:46:05] [290759:00046fc8] [DEBUG][com.freerdp.primitives] - [primitives_autodetect_best]: primitives autodetect, using generic
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_set_negotiation_enabled]: Enabling security layer negotiation: TRUE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_set_restricted_admin_mode_required]: Enabling restricted admin mode: FALSE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_enable_rdp]: Enabling RDP security: TRUE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_enable_tls]: Enabling TLS security: TRUE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_enable_nla]: Enabling NLA security: TRUE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_enable_ext]: Enabling NLA extended security: FALSE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_enable_rdstls]: Enabling RDSTLS security: FALSE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_enable_aad]: Enabling RDS AAD security: FALSE
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.rdp] - [rdp_client_transition_to_state][0xfa03c60]: CONNECTION_STATE_INITIAL --> CONNECTION_STATE_NEGO
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core] - [freerdp_tcp_is_hostname_resolvable]: resetting error state
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core] - [freerdp_tcp_default_connect]: resetting error state
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core] - [freerdp_tcp_default_connect]: connecting to peer 127.0.0.1
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_connect]: state: NEGO_STATE_NLA
[21:36:46:29] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_attempt_nla]: Attempting NLA security
[21:36:46:30] [290759:00046fc8] [DEBUG][com.freerdp.core.nego] - [nego_send_negotiation_request]: RequestedProtocols: 3
[21:36:46:223] [290759:00046fc8] [ERROR][com.freerdp.core.transport] - [transport_read_layer]: BIO_read returned a system error 104: Connection reset by peer
[21:36:46:223] [290759:00046fc8] [ERROR][com.freerdp.core] - [transport_read_layer]: ERRCONNECT_CONNECT_TRANSPORT_FAILED [0x0002000D]
[21:36:46:224] [290759:00046fc8] [DEBUG][com.freerdp.core.rdp] - [rdp_finalize_reset_flags][0xfa03c60]: [CONNECTION_STATE_NEGO] reset finalize_sc_pdus
[21:36:46:224] [290759:00046fc8] [DEBUG][com.freerdp.core.rdp] - [rdp_client_transition_to_state][0xfa03c60]: CONNECTION_STATE_NEGO --> CONNECTION_STATE_INITIAL
```

AkechiShiro commented:
Seems to maybe be an issue specific to Podman, will try `docker` right now : https://github.com/dockur/windows/issues/679

AkechiShiro commented:
Seems RDP works with Docker but there are some issue with the setup (exception on line 151 on the wrapped binary or script ??), I choose to setup :
- `current user`
- `manual`
- `skip all official application`

Running `winapps` showed the error tooltip on KDE about the missing application.
![image](https://github.com/user-attachments/assets/1c10e4f5-45b5-45d6-b2bb-ff15a3b06a88)


I did try also the `winapps-launcher` : 
![image](https://github.com/user-attachments/assets/029d415c-f1dc-4d5c-bc33-45e19563306b)

oskardotglobal commented:
@AkechiShiro It is very odd that winapps does not output or print anything, might be an actual issue and not just with the Nix version. There is no usage printed for `winapps` either, it expects a first argument (either `manual <executable path>`, `windows` for the full windows session or simply an `<app name>`)

AkechiShiro commented:
So what do we do ? Do you want to merge this and work on the issues later ? Should I open an issue to note down that `winapps` does not output any `usage` instructions ?

AkechiShiro commented:
Btw @oskardotglobal it would be cool if we could find a way to integrate the use of this https://git.m-labs.hk/M-Labs/wfvm but I'm not sure that the project is actually maintained but it would allow us to declare a Windows VM

MichielMag commented:
> Btw @oskardotglobal it would be cool if we could find a way to integrate the use of this https://git.m-labs.hk/M-Labs/wfvm but I'm not sure that the project is actually maintained but it would allow us to declare a Windows VM
> 
> Maybe we don't need to integrate with it, but we could write documentation if it works well with Nix and Winapps.

I like this idea, but I don't think that repo would be something to use. For instance, I don't understand the decision to use anything other than WinGet or even Scoop.
Also, the code is quite messy in regards of all his building XML with nix code in between...
I'm actually looking into doing something like this, and was looking at Quickemu, and try to use it with autounattend and a couple of winget scripts.

oskardotglobal commented:
I've tested wfvm and didn't find it to be very useful.

oskardotglobal commented:
@AkechiShiro 

> So what do we do ?

I would say we prioritize merging #240 and then I close this PR and look into getting this into Nixpkgs unstable.

oskardotglobal commented:
The only problem would be how to version / update it on Nixpkgs, I know there is some sort of update process but I currently don't know how

oskardotglobal commented:
We could also keep this as a flake and then have the flake somehow get it's current commit, passing that onto the packages. Then winapps could be updated by the user by simply updating the lock

AkechiShiro commented:
@oskardotglobal we may need to tag versions, there is an update mechanism but we can pin commit hash in revision

We can also use unstable in the version, you need to check the convention in contributing on the Nixpkgs repo

oskardotglobal committed:
Merge branch 'main' into feat-nix-packaging

Alper-Celik commented:
> @oskardotglobal we may need to tag versions, there is an update mechanism but we can pin commit hash in revision
> 
> We can also use unstable in the version, you need to check the convention in contributing on the Nixpkgs repo

you might want to look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#versioning nixpkgs should support packages with no releases yet

btw thanks for packaging to nix. i will try to setup in a few days

AkechiShiro commented:
Also I prefer that we also offer non flakes option as flakes are still not officially stable @oskardotglobal people could pin using npins or other non flakes pinning tools, such as niv

Alper-Celik commented:
nixpkgs doesn't accept flakes as they accept package function that uses same version of dependencies of nixpkgs. but if you want to distribute to nix users outside of nixpkgs and having control over your nixpkgs rev you can make your flake usable to non flake users with https://github.com/edolstra/flake-compat for mostly simple packages and https://github.com/ursi/get-flake for exposing whole flake in non flake nviroments

bbenno commented on README.md:
```diff
@@ -463,6 +463,70 @@ The installer can be run multiple times. To update your installation of WinApps:
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+
+    winapps = {
+      url = ""github:winapps-org/winapps;
```
```diff
- url = ""github:winapps-org/winapps;
+ url = "github:winapps-org/winapps";
```

AkechiShiro replied:
Thanks for your review I completely missed this, I built the package without using the flakes feature.

CHN-beta committed:
use netcat instead of netcat-gnu

oskardotglobal committed:
Merge pull request #281 from CHN-beta/feat-nix-packaging

use netcat instead of netcat-gnu

oskardotglobal committed:
fix: typo

oskardotglobal commented:
My biggest problem currently is how do we keep the flake's commit hash (the `rev` variable) up to date with the latest commit from main?

Alper-Celik commented:
> My biggest problem currently is how do we keep the flake's commit hash (the `rev` variable) up to date with the latest commit from main?

İ think it might be possible to create derivation that outputs rev of main from source but nix might not copy .git folder to derivation

AkechiShiro commented on README.md:
```diff
@@ -463,6 +463,70 @@ The installer can be run multiple times. To update your installation of WinApps:
+
+### On NixOS without Flakes
+
+[Flakes aren't real and they can't hurt you.](https://jade.fyi/blog/flakes-arent-real/).
```
I think we can propose a better solution, if the 2 packages are accepted into Nixpkgs upstream

scraptux commented:
I think NixOS implementation should include options to enable it as a service. The whole point of NixOS is to configure your system declaratively - adding winapps as is to nixpkgs and then having to run it again to configure it would be against the idea of NixOS.

Especially since stuff like impermanence would make using this package harder than just adding .desktop or .rdp files.

oskardotglobal commented:
I thought about implementing a service (or more realistically a program) but then ideally it would create the VM too and that was a tad too complicated to me. Feel free to contribute!

scraptux commented:
I had something similar planned for personal usage. I also believe creating a complete VM and declaratively configuring it (especially windows) is not worth it. I had planned to point the service to an already created vm-image. Creation of this image could be done by a small script.

I'd gladly try to spin something like this up - however won't be able to until end of next month.

Maybe somebody else has a better idea to integrate into NixOS?

Alper-Celik commented:
> I think NixOS implementation should include options to enable it as a service. The whole point of NixOS is to configure your system declaratively - adding winapps as is to nixpkgs and then having to run it again to configure it would be against the idea of NixOS. Especially since stuff like impermanence would make using this package on NixOS harder than just adding .desktop or .rdp files.
> 
> For users on nix and not NixOS this, of course, helps already.

this pr doesn't seem to be adding nixos modules.
nixos module could be added but i think it would be better as separate pr

oskardotglobal committed:
feat: flake-compat

oskardotglobal committed:
Merge branch 'feat-nix-packaging' of github.com:winapps-org/winapps into feat-nix-packaging

oskardotglobal committed:
feat: update packages

oskardotglobal committed:
feat: auto-update flake weekly

oskardotglobal committed:
feat: properly set versions

AkechiShiro commented on packages/winapps-launcher/default.nix:
```diff
@@ -0,0 +1,77 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/winapps-org/WinApps-Launcher";
+    description = "Graphical launcher for WinApps. Run Windows applications (including Microsoft 365 and Adobe Creative Cloud) on GNU/Linux with KDE, GNOME or XFCE, integrated seamlessly as if they were native to the OS. Wayland is currently unsupported.";
```
I tested under KDE Plasma 6 Wayland, however it did automatically use xWayland I presume hence worked out of the box, maybe we should try on GNOME Wayland as well and one last that's heavily used, I'm not sure which, Sway may definitely have issues, Hyprland no clue, never used before.

AkechiShiro commented on packages/winapps/setup.patch:
```diff
@@ -0,0 +1,90 @@
+-  ${COMMAND_TEXT}./setup.sh --help${CLEAR_TEXT}                                        # Display this usage message."
++  ${COMMAND_TEXT}winapps-setup --user${CLEAR_TEXT}                                        # Install WinApps and selected applications in ${HOME}
++  ${COMMAND_TEXT}winapps-setup --system${CLEAR_TEXT}                                      # Install WinApps and selected applications in /usr
++  ${COMMAND_TEXT}winapps-setup --user --setupAllOfficiallySupportedApps${CLEAR_TEXT}      # Install WinApps and all officially supported applications in ${HOME}
++  ${COMMAND_TEXT}winapps-setup --system --setupAllOfficiallySupportedApps${CLEAR_TEXT}    # Install WinApps and all officially supported applications in /usr
```
Ideally we shouldn't have any installations commands (booting into a prior generation on NixOS, what would happen if Winapps is not in the store ? But has installed files, pretty impure).

In an ideal rewrite we would declare what the user wants to have as officially supported in the service or override attributes (all of them should be set to null or false, aka I just want to use cmd.exe for instance, no official apps), but the PR has been worked on for awhile and we probably need to address this issue later.

oskardotglobal replied:
I think these are all things better addressed in a rewrite. I'm currently working on it

oskardotglobal replied:
If you think it's okay to merge, go ahead

AkechiShiro replied:
I would like someone to test it (or I might next week), if you did test it, please let me know.

I left the feedback based on code review alone, I believe there should be tests done, 

Last time there was the "usage" strings, that seemed to be missing, which was confusing for UX + used Podman which made me waste a few hours, unable to RDP.

Alper-Celik replied:
i did test it a while back on your branch with libvirt backend should be this rev: e8dc55a83abd3eb6b275990fb19bdb589b6ce13d

issues i had was:
- `winapps-setup` sometimes didnt worked but worked after a few tries
- and the genareted scripts depended on `/home/alper/.local/bin/winapps` which wasnt there so i symlinked like this `ln -s /run/current-system/sw/bin/winapps /home/alper/.local/bin/winapps` and it worked

Alper-Celik replied:
i will probably test a bit more with lastest version of it tomorrow

oskardotglobal replied:
@Alper-Celik I've gone ahead and addressed those issues in the latest commit, I've lost those changes when transitioning from `sed` commands to patch files.

AkechiShiro commented on packages/winapps/winapps.patch:
```diff
@@ -0,0 +1,11 @@
+     local USER_GROUPS=$(groups "$(whoami)")
+
+-    if ! (echo "$USER_GROUPS" | grep -q -E "\blibvirt\b") || ! (echo "$USER_GROUPS" | grep -q -E "\bkvm\b"); then
++    if ! (echo "$USER_GROUPS" | grep -q -E "\blibvirtd\b") || ! (echo "$USER_GROUPS" | grep -q -E "\bkvm\b"); then
```
I think this also could be improved, using an assertion to check if the group exists and if the user evaluating the flake is inside, but I need to check if there are any modules in NixOS doing these kind of assertions or even packages (I haven't seen any AFAIK)

It would fail that way :
`trace: <username> not in group kvm and libvirtd, please check <doc-to-declare groups member-groups-wiki>`

AkechiShiro reviewed (approved):
I think the PR is in a good place, there are still improvements points but we need to iterate, if we have more users, we'll get more feedback and we'll be able to polish rough edges.

I haven't found time to test it, I might be able to in a week or so.

I recommend we also **note that docker** is recommended for now, a workaround for podman was mentioned (RDP does not work) but not tested yet.

For any licensing issues (ARR related), I think we could distribute the package licensed as unfree (if there is any proprietary blobs) if we can separate the proprietary files, we could also bundle them in a standalone package under license.unfree (`winapps-proprietary`), but even if it's denied upstream at least, we could add hints in the description that the bundle is missing, link towards it, or maybe an attribute (override by user), 

```nix
{withBundle = true;} # license.unfree;
{WithBundle = false;} # license.agpl3plus;
```

By default, we'd set `withBundle` to `false`.

Thanks a lot for the work done @oskardotglobal and everyone else for the feedback.

oskardotglobal committed:
fix: iron out some more issues

This addresses the issues in https://github.com/winapps-org/winapps/pull/234#discussion_r1813582112
and closes #299.

oskardotglobal committed:
fix: don't change mode for owner but for current user

oskardotglobal commented:
Since #299 and #300 are resolved now, this is OK to merge

oskardotglobal commented:
The proposed NixOS module and getting this into nixpkgs should be discussed in a new discussion created for this purpose.
